Implement Shadermodel 6.6 bindless feedback and display it to the pipeline state view

bdumesnil opened the pull request:
<!--
Before submitting a pull request you are strongly recommended to read the
docs/CONTRIBUTING.md file which gives some information on how to prepare a
change:

https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING.md

For small changes you don't have to read the document end to end, but should at
least look at the sections on how to ensure your code and commits are formatted
according to the style requirements.
-->

## Description

This pull request will add support for shader model 6.6 to the bindless feedback and the pipeline state view.
It also fixes a bug in the current implementation of bindless feedback caused by an integer overflow and update the dx.op.annotateHandle instruction's disassembly to match the current DXC implementation.


Here you can see the before/after of updating dx.op.annotateHandle in the disassembly : 
![6_6_annotate](https://user-images.githubusercontent.com/3042591/152869086-57402b52-979f-4134-8841-004b31e72af4.jpg)


Here you can see the before/after of fixing the integer overflow using shadermodel 6.5 bindless : 
![6_5_overflow](https://user-images.githubusercontent.com/3042591/152869102-5fa71024-f3df-4ae9-aa19-3943e43e4b8d.jpg)


Here you can see the before/after of supporting classic and new bindless feedback in shadermodel 6.6 : 
![6_6_table](https://user-images.githubusercontent.com/3042591/152869129-fedc23e4-a012-46ae-9231-6c74dadaaf04.jpg)
![6_6_heap](https://user-images.githubusercontent.com/3042591/152869148-1e959a9d-4612-4680-9ec8-b8318bac5964.jpg)



### Implementation notes : 

* d3d12_shader_feedback.cpp : 
	Handling classic binding for shader model 6.6 basically consists in using createHandleFromBinding instead of createHandle
	
	To handle direct heap access I added one special D3D12FeedbackKey for resources and one for samplers for each shader stages that need them, then handle createHandleFromHeap when patching the shaders.
	After replaying the command, I iterate over those D3D12FeedbackKey and create a D3D12FeedbackBindIdentifier for each resource accessed.
	
	Since the disassembly doesn't give a way to get the name of the variable linked to a createHandleFromHeap, I only create one D3D12FeedbackKey slots per shader stage for all createHandleFromHeap instances instead of one slot per createHandleFromHeap instance ( contrary to createHandle/createHandleFromBinding ).
	If there was a way to to get the variable name ( maybe from the debug info ? ), this could easily be modified to add one D3D12FeedbackKey slot per createHandleFromHeap instances.
	

* d3d12_replay.cpp :
	I iterate over all the created D3D12FeedbackBindIdentifier usages and create/fill the necessary D3D12Pipe::RootSignatureRange.
	
* D3D12PipelineStateViewer.cpp :
	Mostly checking if the resources are directHeapAccess to avoid printing Space/Register that don't make sense in this case in the "Pipeline State" view
	
* dxil_disassemble.cpp
	When shadermodel 6.6 was added to DXC the arguments of the dx.op.annotateHandle instruction changed, which could lead to access to non-existent arguments and crash.
	I updated the disassembly to string function to reflect those changes.

Commit:
Fix dx.op.annotateHandle having changed format.

Commit:
Fix possible uint32 overflow when using unbounded resource tables.

Commit:
Add support for shader model 6.6 bindless resources.

baldurk commented on qrenderdoc/Windows/PipelineState/D3D12PipelineStateViewer.cpp:
```diff
@@ -1241,19 +1272,26 @@ void D3D12PipelineStateViewer::setShaderState(
+            spaceStr = tr("");
+          }
+          else if(rootElements[i].immediate)
+            rootel = tr("#%1 Static").arg(rootElements[i].rootSignatureIndex);
```
Minor style nitpick - please 'balance' the braces so that if the `if()` has braces, add braces to all matching `else` branches.

baldurk commented on qrenderdoc/Windows/PipelineState/D3D12PipelineStateViewer.cpp:
```diff
@@ -1397,12 +1439,16 @@ void D3D12PipelineStateViewer::setShaderState(
-            if(bind != NULL && bind->arraySize > 1)
-              regname += tr("[%1]").arg(b.bind - bind->bind);
+            if(directHeapAccess)
+              regname = tr("");
```
FYI same thing here - balance out the braces.

baldurk commented on renderdoc/driver/d3d12/d3d12_replay.cpp:
```diff
@@ -1355,6 +1358,75 @@ void D3D12Replay::FillRootElements(uint32_t eventId, const D3D12RenderState::Roo
+        const D3D12_SAMPLER_DESC &sampDesc = desc->GetSampler();
+        FillSampler(samp, sampDesc);
+      }
+      if(curUsage->bindType == BindType::ConstantBuffer)
```
I think this should be an `else if`? Otherwise samplers will fill the sampler properties and then also the resource view properties.

bdumesnil replied:
Indeed it should be an ``else if``

baldurk commented on renderdoc/driver/d3d12/d3d12_replay.cpp:
```diff
@@ -1355,6 +1358,75 @@ void D3D12Replay::FillRootElements(uint32_t eventId, const D3D12RenderState::Roo
+      {
+        D3D12Descriptor *desc =
+            (D3D12Descriptor *)resourceHeap->GetCPUDescriptorHandleForHeapStart().ptr;
+        desc += curUsage->descIndex;
```
I think it would be good to set `view.tableIndex` (and similarly for samplers & cbuffers) to `curUsage->descIndex` here and display that in the pipeline viewer as a heap index. We don't have much information to provide the user unfortunately with this indexing scheme, so we should display what we can. It might be helpful for people trying to correlate to shader indices they see in the data.

bdumesnil replied:
Good idea, I will add it.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -72,6 +72,17 @@ struct D3D12FeedbackSlot
 static const uint32_t numReservedSlots = 4;
 static const uint32_t magicFeedbackValue = 0xbeebf33d;
 
+static D3D12FeedbackKey GetDirectHeapAccesKey(DXBCBytecode::OperandType opType)
```
Minor typo here, `Acces` instead of `Access`. It's also in a couple of places below, I won't spam a comment on every one but please double check.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -892,23 +1222,32 @@ void D3D12Replay::FetchShaderFeedback(uint32_t eventId)
   }
 
-  bytebuf editedBlob[5];
+  bytebuf editedBlob[6];
```
Can you explain the reason for this change? It's currently 5 entries since either we're patching compute, or graphics, but not both at the same time. So either compute is in [0] or the graphics stages are in [0...5]. It's not a big deal to change either way but I'm wondering why it was made.

If there's a reason to have compute and graphics at the same time, compute should be in index [5] since that's consistent with the convention elsewhere in the codebase and this won't cause any surprises for future maintenance.

bdumesnil replied:
The idea is to have one index for each stage to simplify indexing farther in the function when adding direct access usages.

You are right, it would be better to use the same value as other parts of the codebase, I think it would be better to replace the hardcoded values and use the enum ShaderStage instead.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -1218,6 +1569,55 @@ void D3D12Replay::FetchShaderFeedback(uint32_t eventId)
+                if(desc->GetType() == D3D12DescriptorType::SRV)
+                  directAccessIdentifier.bindType = BindType::ReadOnlyResource;
+                if(desc->GetType() == D3D12DescriptorType::UAV)
+                  directAccessIdentifier.bindType = BindType::ReadWriteResource;
```
If I understand your intention right, reading this descriptor type here from the heap is so that resource identifiers can be properly sorted by type below, so that when this is iterated in `D3D12Replay::FillRootElements` it encounters all the types in order contiguously even if they're interleaved in the heap, and you only get a couple of root element ranges?

To be clear this works just fine, but I'd rather not be reading the current contents of the resource heap here in the feedback fetch. I'd rather keep that separate as much as possible and only deal with the shader's view and information here.

I think instead we should use the `dx.op.annotateHandle` type information that we have. That tells us for each `createHandleFromHeap` what type the shader is reading it as, which is the closest equivalent to traditional bindings where we push this information back based on shader metadata/access. This means instead of just a magic uint32 value to indicate whether a slot is used, we probably want to reserve some bits for a type indicator. A mismatch to the actual contents of the descriptor heap is illegal, but we could detect that if we want. If we reserve enough bits (we really don't need more than 1 for the magic value) we could also detect multiple use of the same heap index via incompatible types by seeing if multiple bits are set.

The only issue is that I'm not clear on whether `dx.op.annotateHandle` is actually required or if it's considered debug information, the docs are of course non-existent or unclear. However even if it's not, it would be better to pass these back 'typeless' and let the root element filling code sort on the other end based on the heap contents at that point.

bdumesnil replied:
It is indeed to get the resource identifiers properly sorted.
I originally started by using ``dx.op.annotateHandle`` but as you said the lack of documentation made me fear that I could miss some cases, however from my limited testing the instruction seems to be there even compiled as -O3 without debug informations.

I will put it back instead of looking at the heap, though rather than using bits in the feeback value I'd rather allocate a range in the feedback UAV per resource type ( [SRV][UAV][CBV][SAMPLER] instead of current [SRV,UAV,CBV][SAMPLER] ranges ), it may use more memory but since the ideal case would be a range per ``CreateHandleFromHeap`` to correlate resource use with shader source, I think it it non-significant.

baldurk replied:
Yes I also don't think we should hardcode an assumption for `dx.op.annotateHandle`, since there seems no strong guarantee that it's always present. Unfortunately testing ad-hoc means we can make it our primary path but we can't assume it will be there. If it's missing I think it's fine to return in unsorted order, or not generate results entirely, and handle that case if it comes up, as long as the code doesn't crash.

I don't think there's any reason to allocate more feedback space right now. Using a more intelligent identifier should work as we have plenty of space still. You're right that we could allocate more feedback space if needed, but I don't think we should do that without good reason. If in future we needed to correlate with shader source we would either still have free bits to encode a per-descriptor access ID, or else if we wanted to track accesses to the same descriptor from multiple source accesses we could move to a system like vulkan's shader printf where instead of fixed output slots we append to a buffer with each access. I think at that point having a feedback range per source access per type is not so feasible.

bdumesnil replied:
Ok I'll change it to use bits in the magic value.
Would you also merge the feedback range for samplers with the range for SRV/CBV/UAV ?
So that an index in the feedback range could represent at the same a descriptor in the sampler heap and one of SRV/CBV/UAV descriptor in the resource heap

baldurk replied:
Sampler heaps are typically a lot smaller (you don't tend to get 1 million entry heaps there) but I think you could do that. I don't mind if it's still separate or unified.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -187,6 +198,62 @@ static bool AnnotateDXBCShader(const DXBC::DXBCContainer *dxbc, uint32_t space,
+{
+  resourcesDirectAccess = 0;
+  samplersDirectAcces = 0;
+  DXIL::ProgramEditor editor(dxbc, 0, editedBlob);
```
I'd rather not construct the `DXIL::ProgramEditor` here. DXIL is an absolutely _awful_ format and processing it is not fast, especially so if we're setting up for editing which we don't need here.

I think instead it would be better to cache this information when doing the original `DXIL::Program` process (which we do once per shader for reflection) and store it as metadata. You can access it via `DXBC::DXBCContainer::GetDXILByteCode()`

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -187,6 +198,62 @@ static bool AnnotateDXBCShader(const DXBC::DXBCContainer *dxbc, uint32_t space,
+    return;
+  bool isShaderModel6_6OrAbove =
+      dxbc->m_Version.Major > 6 || (dxbc->m_Version.Major == 6 && dxbc->m_Version.Minor >= 6);
+  RDCASSERTEQUAL(isShaderModel6_6OrAbove, true);
```
This isn't a problem, but just FYI there's no need to do explicit equals checks here when the test or value is just boolean. The main idea of `RDCASSERTEQUAL` is so you can assert two values and have the assert automatically print out their values when it fails. In this case if the assert fails we know by definition what the value is :smile: so `RDCASSERT(isShaderModel6_6OrAbove)` would be fine. You don't have to change that though if you don't want to.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -195,16 +262,64 @@ static bool AnnotateDXILShader(const DXBC::DXBCContainer *dxbc, uint32_t space,
+
+    DXIL::Type funcPtrType;
+    funcPtrType.type = DXIL::Type::Pointer;
+    funcPtrType.inner = editor.AddType(funcType);
```
Currently editor functions like `AddType` or `DeclareFunction` do _not_ do any deduplication, so this could create duplicate types. With DXIL it's likely impossible to say if this is safe or not, so I think your code needs to check to see if the type exists already here, similar to what you do with attribute sets or the `dx.types.ResBind` type itself.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -633,41 +810,94 @@ static bool AnnotateDXILShader(const DXBC::DXBCContainer *dxbc, uint32_t space,
+          RDCERR("Unexpected non-constant argument to createHandleFromBinding");
+          continue;
+        }
+        // resbind can sometime be empty
```
Can you give more information about when this would be empty? I don't understand what the instruction would mean in that case.

bdumesnil replied:
Looking at it in more details, it seems this case is handled incorrectly, resbind is empty when the variable is nullconst, so it should be handled as "all args to 0" instead of being skipped

baldurk replied:
Ahh, that makes sense as the dxil reader doesn't currently 'expand' null constants of aggregate type like that. In future that might be something to consider, but dxil is absolutely awful to work with so better to handle it deliberately here exactly as you say.

baldurk reviewed (changes requested):
Thanks for this change! It looks good overall and I think it's the right direction for now. I've left some comments about a few things that should be addressed.

One thing that came up for me in testing, I only have one very simple SM6.6 demo that someone shared with me, but it doesn't use any traditional bindings at all which means the `dx.resources` metadata entry is completely missing. This causes `AnnotateDXILShader` to early out because it's missing, but it also means that when declaring our feedback UAV that metadata will need to be added.

It would be very nice if you could add a simple auto-test for this :grinning:. Especially for simple cases like each type, shaders with/without traditional bindings. But it's not required by any means.

----

Thoughts looking forward (none of this is feedback for the PR, but if you are interested in working on this further this would be relevant):

The main downside with this - which I don't think you can fix - is that we completely lose reflection information for these accesses. Looking at the issues on dxc this was raised a long time ago so I don't have much hope that it will change: https://github.com/microsoft/DirectXShaderCompiler/issues/3531 

Most importantly any constant buffer accessed this way has no formatting information so we can't do any better than displaying as `uint[]` and letting the user set the format. This is a big loss in usability but I don't immediately see any way around it. Unfortunately with DXIL there's no way to provide proper side-band debug information of our own.

The best we could do is use the line number information to search for the original line, and look for a specially formatted comment. E.g. if the user did:

```
// RENDERDOC: type=MyConstBuffer name=frameConstants
ConstantBuffer<MyConstBuffer> frameConstants= ResourceDescriptorHeap[idx];
```

Then we could read this back and use that to reconstruct the missing information. However I suspect that people wouldn't be willing to put that kind of annotation in their shaders. Especially when it duplicates information that the compiler should be able to put into debug info.

Also on a different angle, while I think this is as clean a way as possible to integrate this feedback into the current pipeline state & binding representation, it probably no longer makes sense to represent bindings in this way. For bindless array access it still sort of made sense to let static information define the layout, but for this type of direct descriptor loads we probably want to move accesses from the shader feedback into a separate path. However that will have to wait until things are bit more clear with how this will work in the general case (especially depending on how debug info is resolved).

bdumesnil commented:
Thank you for the feedbacks, I'll go back and fix them.
I agree that losing the reflection informations is a big downside, both for debugging tools and contentpipelines.

For the auto-tests, do you have an example of similar tests in the codebase that I could look at ?

baldurk commented:
The closest auto-test at the moment would be `D3D12_Descriptor_Indexing` which tests to make sure the existing bindless feedback works with array bindings. The source of the C++ demo program is [here](https://github.com/baldurk/renderdoc/blob/v1.x/util/test/demos/d3d12/d3d12_descriptor_indexing.cpp) and the python script which tests it is [here](https://github.com/baldurk/renderdoc/blob/v1.x/util/test/tests/D3D12/D3D12_Descriptor_Indexing.py).

The C++ program sets up a simple case where arrayed descriptors are used sparsely, and the python script uses the same replay API as you've used to expose this information to the UI, to check on a capture of the demo and make sure the proper bindings are reported as used.

As I mentioned you don't have to add a test for this, but it would definitely be appreciated :).

Commit:
Update D3D headers in test project to latest D3D12 headers

Commit:
Fix test of D3D12 bindless feedback.

Fix a typo in a comment in D3D12_Descriptor_Indexing.py.
Fix UAV using a typed buffer view instead of a structured buffer view.
Fix SetDescriptorHeaps being called before ClearUnorderedAccessViewUint

Commit:
Add test of D3D12 bindless feedback for shadermodel 6.6.

bdumesnil commented:
I amended my last commit to make all the above changes, it also include a fix for a crash introduced in the previous commit when selecting a bindless constantbuffer with shadermodel 6.6 direct heap access.

On top of that, I added commits with an update to the D3D12_Descriptor_Indexing test for shader model 6.6 table and heap bindless techniques.
It includes two fixes to the existing test ( wrong format set on an UAV view making the test fail and an error reported by the d3d debuglayer because the descriptor heap is set too late ).
I had to update the d3d12 headers used by the demo to get access to the new D3D flags ( D3D12_ROOT_SIGNATURE_FLAG_SAMPLER_HEAP_DIRECTLY_INDEXED and D3D12_ROOT_SIGNATURE_FLAG_CBV_SRV_UAV_HEAP_DIRECTLY_INDEXED ).

baldurk commented on util/test/demos/d3d12/d3d12_descriptor_indexing.cpp:
```diff
@@ -255,9 +255,9 @@ float4 main(v2f IN) : SV_Target0
+        MakeUAV(structBuf).Format(DXGI_FORMAT_R32G32_UINT).CreateGPU(15);
     D3D12_CPU_DESCRIPTOR_HANDLE structCPU =
-        MakeUAV(structBuf).Format(DXGI_FORMAT_R32_UINT).CreateClearCPU(15);
+        MakeUAV(structBuf).Format(DXGI_FORMAT_R32G32_UINT).CreateClearCPU(15);
```
Why is this format change needed? These uav handles are only used for clearing, and it should be entirely safe to clear as plain uints to zero. I don't think D3D has any rules about referring to the same buffer with different UAV formats, and in the general case for a complex struct there won't be a matching format (though in this case there is).

I don't get any D3D debug warnings but I also don't get one about SetDescriptorHeaps below, which doesn't surprise me too much as the D3D debug layer is quite unreliable.

bdumesnil replied:
Yes this specific format could be left as DXGI_FORMAT_R32_UINT since CreateClearCPU will automatically adjust the number of elements to be doubled ( at half the real size ).

However the previous line : 
> MakeUAV(structBuf).Format(DXGI_FORMAT_R32G32_UINT).CreateGPU(15);

need to be changed because it is used by ClearUnorderedAccessViewUint and for writing/reading the tex_ref between the compute shader and pixel shader.

I mainly changed both to keep them coherent, I don't think ClearUnorderedAccessViewUint impose both descriptors to be identical.

baldurk replied:
Ahh I see what you mean. Then actually a different fix is needed - it's not valid to use a typed buffer (like R32G32_UINT) with a structured buffer binding. We need one descriptor that's typed for clearing (since you can't clear with a structured descriptor), which can still just be R32_UINT, and a different descriptor that's structured with `.Structured(8)` like the SRV.

baldurk commented on util/test/demos/d3d12/d3d12_descriptor_indexing.cpp:
```diff
@@ -161,21 +249,29 @@ float4 main(v2f IN) : SV_Target0
-    ID3D12PipelineStatePtr computepso[2];
+    graphicssigs[3] = MakeSig({}, D3D12_ROOT_SIGNATURE_FLAG_ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT |
+                                      D3D12_ROOT_SIGNATURE_FLAG_SAMPLER_HEAP_DIRECTLY_INDEXED |
+                                      D3D12_ROOT_SIGNATURE_FLAG_CBV_SRV_UAV_HEAP_DIRECTLY_INDEXED,
```
This code (and the 6.6 tests in general) will need to be made optional/conditional based on shader model support at runtime, since this functionality will only be available if people are using the new D3D distributable SDK with SM6.6.

baldurk commented on renderdoc/driver/d3d12/d3d12_shader_feedback.cpp:
```diff
@@ -284,7 +362,14 @@ static bool AnnotateDXILShader(const DXBC::DXBCContainer *dxbc, uint32_t space,
+      DXIL::Metadata tmpResources;
+      tmpResources.children.push_back(editor.AddMetadata(tmpResList));
+      resources = editor.AddMetadata(tmpResources);
+    }
```
This fixes the crash I was seeing during patching when `dx.resources` is missing, but it fails to create a pipeline state because the dx.resources named metadata doesn't exist. You'll need to add a new `AddNamedMetadata` I think here, but otherwise I _think_ it shouldn't be complicated.

bdumesnil replied:
You are right, strangely enough this only fail on one of my two PCs, adding a AddNamedMetadata does fix the fail.

baldurk reviewed (changes requested):
Thanks for this! These changes look good now, I've made a couple of notes about some issues still.

I tried to run the test but it fails for me not too surprisingly because the test program isn't set up to use the new D3D12 SDK that adds support or SM6.6 and heap indexing. How are you running it? I'll need to see if it's worth adding some command line option to load the new SDK, since that would probably mean porting over some hooking code into the demos project.

One minor note - updating the D3D12 headers in the demos project is fine (I usually try and update them in lockstep but I sometimes forget) but please use the same version as in the main driver backend, so they don't get out of sync. The one there should be new enough for the flags you want.

bdumesnil commented:
It seems I have the required DX12 SDK included on my windows ( Win10 21H2 and Win11 ).
I'll add the runtime check for sm6.6 support and make sure the D3D headers are identicals.

baldurk commented:
Ah I didn't realise that stuff was available by default on newer windows versions. For what it's worth my win10 test machine (my primary machine is win7) is version 1909, I'm not sure what that corresponds to since windows versions are unnecessarily confusing.

baldurk reviewed (approved):
This looks good, I'm going to merge :+1:

Just for clarity I'm going to make a couple of follow up commits for things that are not your fault so I won't ask you to address them in this PR:

Since you started I changed the DXIL representation so that function types aren't `ptr-to-function` but just `function` (see 1591ebcb1e211dc6a864650feda959fede6d2514). This is not the LLVM convention but some DXIL emitted by mesa's custom emitter does this. It just amounts to removing a pointer indirect from your code, so no big deal.

I also tested on the SM6.6 demo I had and got a crash when selecting the draw, but digging deeper it is using `SamplerDescriptorHeap` even though it doesn't actually bind a sampler heap - it uses static samplers. I'm not sure how this works and I can't imagine it's intended, the D3D debug layers don't complain but that doesn't mean much as there are many basic obvious errors they don't catch. Since this SM6.6 indexing seems quite easy to make errors with and the debug layers don't catch it I'm going to add a quick check to see if the heap is valid to avoid the crash, even though normally I don't add validation protection in RenderDoc.

bdumesnil commented:
Great ! Thanks for your time reviewing.
